### PR TITLE
HPCC-17973 Use a read-only common value table for attributes.

### PR DESCRIPTION
**Pull request by jakesmith**
Similar to approach taken for attribute names, use a pre-populated
read-only table for names.
The table is pre-populated with values derived from environment
startup and regression suite runs.
It is also pre-populated with unsigned <10000 as many of them are
are very commonly used.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:

Speed and memory tests performed with a load/serialize/copy of large xml file (details in JIRA)
Full regression run.

### CODE REVIEW

**Comment by jakesmith** [comment]
@richardkchapman - please review

**Comment by jakesmith** [comment]
@richardkchapman - trimmed const values.

Please review

File: system/jlib/jptree-attrvalues.hpp [hunk 1]
```diff
@@ -0,0 +1,953 @@
...
+    "* * * * *",
+    "*",
+    "\\,",
+    "0 secs",
+    "-1"
+    "1:1",
+    "1:M",
+    "* 2 * * *",
+    "* 3 * * *",
+    "aaalibrary2",
```
Comment (richardkchapman) [inline system/jlib/jptree-attrvalues.hpp]: Specific to regression suite runs...

File: system/jlib/jptree-attrvalues.hpp [hunk 2]
```diff
@@ -0,0 +1,953 @@
...
+    "Action #1",
+    "Action #2",
+    "active",
+    "activities",
+    "Aggregate",
+    "Alias",
+    "All Denormalize",
+    "All Denormalize Group",
+    "All Join",
+    "anderson",
```
Comment (richardkchapman) [inline system/jlib/jptree-attrvalues.hpp]: specific to regression suite runs

File: system/jlib/jptree-attrvalues.hpp [hunk 3]
```diff
@@ -0,0 +1,953 @@
...
+    "templ",
+    "Temporaries",
+    "temporary",
+    "tempr",
+    "temps",
+    "Text",
+    "textsearch",
+    "TextSearch",
+    "tf01",
+    "tf37",
```
Comment (richardkchapman) [inline system/jlib/jptree-attrvalues.hpp]: this list makes no sense - probably specific to regression suite again

File: system/jlib/jptree-attrvalues.hpp [hunk 4]
```diff
@@ -0,0 +1,953 @@
...
+    "tfV9",
+    "thor",
+    "Thor",
+    "thorlib",
+    "Thorlib",
+    "ThorMaster",
+    "Thor process",
+    "ThorQueueAccess",
+    "thor_roxie",
+    "thor_roxie.eclserver",
```
Comment (richardkchapman) [inline system/jlib/jptree-attrvalues.hpp]: depends on names of queues

**Comment by jakesmith** [comment]
@richardkchapman - trimmed some more and added a couple more auto generated sequences.

File: system/jlib/jptree-attrvalues.hpp [hunk 5]
```diff
@@ -0,0 +1,770 @@
...
+    "Application",
+    "archive",
+    "ASSERT",
+    "Associated",
+    "atftpd",
+    "Attr",
+    "Begin Loop",
+    "blank",
+    "blocked",
+    "Blocked by fileservice activity: Copy",
```
Comment (richardkchapman) [inline system/jlib/jptree-attrvalues.hpp]: Dubious

File: system/jlib/jptree-attrvalues.hpp [hunk 6]
```diff
@@ -0,0 +1,770 @@
...
+    "activities",
+    "Aggregate",
+    "Alias",
+    "any",
+    "anyone",
+    "Application",
+    "archive",
+    "ASSERT",
+    "Associated",
+    "atftpd",
```
Comment (richardkchapman) [inline system/jlib/jptree-attrvalues.hpp]: Dubious

File: system/jlib/jptree-attrvalues.hpp [hunk 7]
```diff
@@ -0,0 +1,770 @@
...
+    "copy",
+    "Count",
+    "Count Project",
+    "cpp",
+    "created_by",
+    "created_for",
+    "CRON",
+    "csv",
+    "Csv Write",
+    "Customer of HPCCSystems",
```
Comment (richardkchapman) [inline system/jlib/jptree-attrvalues.hpp]: Dubious

File: system/jlib/jptree-attrvalues.hpp [hunk 8]
```diff
@@ -0,0 +1,770 @@
...
+    "ECL",
+    "eclagent",
+    "EclAgent",
+    "ECLagent",
+    "EclAgentBuild",
+    "eclagent_config.xsd",
+    "EclAgent process",
+    "EclAgentProcess",
+    "eclcc",
+    "eclcc-dalgo-0",
```
Comment (richardkchapman) [inline system/jlib/jptree-attrvalues.hpp]: Ones starting eclcc- are all specific to regression suite I think

File: system/jlib/jptree-attrvalues.hpp [hunk 9]
```diff
@@ -0,0 +1,770 @@
...
+    "Inline Row",
+    "inputs",
+    "_internal",
+    "isokay",
+    "_isSpill",
+    "_isTransformSpill",
+    "Item",
+    "Iterate",
+    "Job",
+    "/JobQueues[1]/Queue[1]",
```
Comment (richardkchapman) [inline system/jlib/jptree-attrvalues.hpp]: not [2] ??

File: system/jlib/jptree.ipp [hunk 10]
```diff
@@ -517,15 +517,45 @@ struct AttrStrUnionWithTable : public AttrStrUnion
...
     static RONameTable *roNameTable;
 };
+
+struct AttrStrUnionWithValueTable : public AttrStrUnionWithTable
+{
+    inline const char *get() const
+    {
+        if (flag==3) // no point in also checking !isPtr() afaics
+            return roValueTable->getIndex(idx2)->str_DO_NOT_USE_DIRECTLY;  // Should probably rename this back now!
+        return AttrStrUnion::get();
```
Comment (richardkchapman) [inline system/jlib/jptree.ipp]: I think this means still ends up inlining short ones

Comment (richardkchapman) [inline system/jlib/jptree.ipp]: Actually I suppose it does not - but will still check whether it has done

File: system/jlib/jptree-attrvalues.hpp (hunk 5 above)
Comment (jakesmith) [inline system/jlib/jptree-attrvalues.hpp]: yes I guess an edge case, even though it ranked highly - probably anomaly of regression suite.
I'll remove.

File: system/jlib/jptree-attrvalues.hpp (hunk 6 above)
Comment (jakesmith) [inline system/jlib/jptree-attrvalues.hpp]: not sure what it is, I'll remove though.

File: system/jlib/jptree-attrvalues.hpp (hunk 7 above)
Comment (jakesmith) [inline system/jlib/jptree-attrvalues.hpp]: I'll remove.

File: system/jlib/jptree.ipp (hunk 10 above)
Comment (jakesmith) [inline system/jlib/jptree.ipp]: @richardkchapman - btw, do you agree with the comment on line 525?
If so applies to AttrStrUnionWithTable::set() too.

File: system/jlib/jptree-attrvalues.hpp (hunk 9 above)
Comment (jakesmith) [inline system/jlib/jptree-attrvalues.hpp]: don't know why wasn't generated, I'll add anyway.

**Comment by ghalliday** [comment]
@jakesmith How much benefit is there in adding numbers 1,000..10,000.  It increases the startup overhead from ~1.2ms to ~4ms on my desktop.

**Comment by ghalliday** [comment]
(An experiment with presizing the hash table reduced that to ~3.2ms)

**Comment by richardkchapman** [comment]
Replaced by #10242

**Pull request closed without merging**